### PR TITLE
Private/gokay/pdfscroll2

### PR DESCRIPTION
**Pull request by gokaysatir**
* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [ ] Code is properly formatted
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 1]
```diff
@@ -5472,12 +5472,144 @@ L.CanvasTileLayer = L.Layer.extend({
...
+	// In file based view, all parts may be drawn, not only selected part.
+	_getTileInfoForFileBasedView: function (currentX, currentY, zoom, partHeightPixels, queue) {
+		var yMin = currentY;
+		var yMax = yMin + this._tileSize;
+
+		var parts = [];
+		for (var i = 0; i < this._parts; i++) {
+			var partMin = i * partHeightPixels;
+			var partMax = (i + 1) * partHeightPixels;
+			if ((yMin >= partMin && yMin <= partMax) || (yMax >= partMin && yMax <= partMax)) {
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CanvasTileLayer.js:5375]: What if one or more part(s) can fit completely in a tile at a low enough zoom ?

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 2]
```diff
@@ -5472,12 +5472,144 @@ L.CanvasTileLayer = L.Layer.extend({
...
+		var zoom = Math.round(this._map.getZoom());
+
+		var ratio = this._tileSize / this._tileHeightTwips;
+		var partHeightPixels = Math.round((this._partHeightTwips + this._spaceBetweenParts) * ratio);
+
+		var topLeft = app.sectionContainer.getDocumentTopLeft();
+		var bottomRight = app.sectionContainer.getDocumentBottomRight();
+
+		topLeft = [topLeft[0] - topLeft[0] % this._tileSize, topLeft[1] - topLeft[1] % this._tileSize];
+		bottomRight = [bottomRight[0] - bottomRight[0] % this._tileSize, bottomRight[1] - bottomRight[1] % this._tileSize];
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CanvasTileLayer.js:5441]: Shouldn't we include the last (w.r.t. x and y) tile(s) too which intersects with the document's current view ?

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 3]
```diff
@@ -5472,12 +5472,144 @@ L.CanvasTileLayer = L.Layer.extend({
...
+					tile = this.createTile(this._wrapCoords(queue[i]), L.bind(this._tileReady, this, queue[i]));
+
+					// save tile in cache
+					this._tiles[key] = {
+						el: tile,
+						wid: 0,
+						coords: queue[i],
+						current: true
+					};
+
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CanvasTileLayer.js:5497]: This section related to tile object vs tile img object, will need to adapt to slurping related changes whereby the tileCache has the resolved image elements and that we don't have to deal with img's src attribute. See commits:
https://github.com/CollaboraOnline/online/commit/663016d5d0a0e7e193d9328d40493fb4e9f18519
https://github.com/CollaboraOnline/online/commit/5e64575ab687e203c38e388a2d165d772e352561

This is also true for co-6-4.

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 4]
```diff
@@ -6023,7 +6155,46 @@ L.CanvasTileLayer = L.Layer.extend({
...
+		coords.part = tileMsgObj.part;
+
+		var tile = this._tiles[coords.key()];
+
+		if (tile) {
+			if (tile._invalidCount > 0) {
+				tile._invalidCount -= 1;
+			}
+
+			tile.el.src = img;
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CanvasTileLayer.js:6082]: The above comment about slurping changes applies here too. (img is the resolved image object not the source string).

**Review by dennisfrancis — COMMENTED** [review]
I've left some comments after reading till the commit "Draw: Enable PDF scrolling for tile section". Thanks!

File: loleaflet/src/layer/tile/CommentSection.ts [hunk 5]
```diff
@@ -71,6 +82,542 @@ class CommentSection {
 
 	}
 
+	private adjustComment (comment: any) {
+		var rectangles, color, viewId;
+		comment.trackchange = false;
+		rectangles = L.PolyUtil.rectanglesToPolygons(L.LOUtil.stringToRectangles(comment.textRange || comment.anchorPos), this.map._docLayer);
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CommentSection.ts]: We are trying to get rid of L.Poly*, not urgent but please use CPoly* in future.

File: loleaflet/src/layer/tile/CommentSection.ts [hunk 6]
```diff
@@ -71,6 +82,542 @@ class CommentSection {
...
+		comment.anchorPos = L.LOUtil.stringToBounds(comment.anchorPos);
+		comment.anchorPix = this.map._docLayer._twipsToPixels(comment.anchorPos.min);
+		viewId = this.map.getViewId(comment.author);
+		color = viewId >= 0 ? L.LOUtil.rgbToHex(this.map.getViewColor(viewId)) : '#43ACE8';
+		if (rectangles.length > 0) {
+			comment.textSelected = L.polygon(rectangles, {
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CommentSection.ts]: Is this going to use the old svg overlay layer and not the canvas after all and having a comments section etc is just to sync stuff with other sections specifically the tiles ?

**Review by dennisfrancis — COMMENTED** [review]
Few more comments after reading more patches.

File: loleaflet/src/layer/tile/CanvasTileLayer.js (hunk 1 above)
Comment (gokaysatir) [reply loleaflet/src/layer/tile/CanvasTileLayer.js:5362]: Then partHeightPixels will be lower too. It is a dynamic variable. But really good catch by the way:) I'll check it again to see if everything is ok. Thanks for all the comments, i am taking them into account, some of them are now in co-6-4, i will send some follow-ups when i'm ready for the details like LPoly-CPoly convertions.

**Pull request merged**